jit-compiler/src/compiler.rs

```diff
@@ -104,28 +110,35 @@ pub fn call_cargo(code: &str) -> Result<(Temp, String), String> {
         .join("target")
         .join("release")
         .join(format!("libpowdr_jit_compiled.{extension}"));
-    Ok((dir, lib_path.to_str().unwrap().to_string()))
+    Ok(PathInTempDir {
+        dir,
+        path: lib_path.to_str().unwrap().to_string(),
+    })
 }
 
-/// Loads the given library and creates funtion pointers for the given symbols.
-pub fn load_library(path: &str, symbols: &[&str]) -> Result<SymbolMap, String> {
-    let c_path = CString::new(path).unwrap();
-    let lib = unsafe { dlopen(c_path.as_ptr(), RTLD_NOW) };
-    if lib.is_null() {
-        return Err(format!("Failed to load library: {path:?}"));
-    }
-    let mut result = HashMap::new();
-    for sym in symbols {
-        let extern_sym = extern_symbol_name(sym);
-        let sym_cstr = CString::new(extern_sym).unwrap();
-        let fun_ptr = unsafe { dlsym(lib, sym_cstr.as_ptr()) };
-        if fun_ptr.is_null() {
-            return Err(format!("Failed to load symbol: {fun_ptr:?}"));
-        }
-        let fun = unsafe { std::mem::transmute::<*mut c_void, fn(u64) -> u64>(fun_ptr) };
-        result.insert(sym.to_string(), fun);
-    }
-    Ok(result)
+/// Loads the given library and creates function pointers for the given symbols.
+pub fn load_library(
+    path: &str,
+    symbols: &[&str],
+) -> Result<HashMap<String, LoadedFunction>, String> {
+    let library = Arc::new(
+        unsafe { libloading::Library::new(path) }
+            .map_err(|e| format!("Error loading library at {path}: {e}"))?,
+    );
+    symbols
+        .iter()
+        .map(|&sym| {
+            let extern_sym = extern_symbol_name(sym);
+            let function =
+                *unsafe { library.get::<unsafe extern "C" fn(u64) -> u64>(extern_sym.as_bytes()) }
```

Comment:
I am not sure I agree the loaded function should be `unsafe`. We just wrote its code and built it ourselves, and its rust code is perfectly safe.